### ISSUE TITLE
When converting to YAML, represent scalars with newlines using "|" instead of "\""

### ISSUE THREAD

**Issue by azatoth** [awslabs/aws-cfn-template-flip#34]
Instead of do a normal "-quote, I suggest instead of using | when representing scalars when they contains newlines

**Comment by stilvoid** [comment]
Can you provide a couple of examples of inputs and desired outputs?

**Comment by azatoth** [comment]
For following JSON
```json
{
  "01_add_instance_to_cluster": {
    "command": {
      "Fn::Join": [
        "",
        [
          "#!/bin/bash\n",
          "echo ECS_CLUSTER=",
          {
            "Ref": "ECSCluster"
          },
          " >> /etc/ecs/ecs.config"
        ]
      ]
    }
  },
  "02_install_ssm_agent": {
    "command": {
      "Fn::Join": [
        "",
        [
          "#!/bin/bash\n",
          "yum -y update\n",
          "yum install -y",
          " ",
          "https://amazon-ssm-eu-west-1.s3.amazonaws.com/latest/linux_amd64/amazon-ssm-agent.rpm"
        ]
      ]
    }
  }
}
```

currently produces (with clean):

```yaml
01_add_instance_to_cluster:
  command: !Sub '#!/bin/bash

    echo ECS_CLUSTER=${ECSCluster} >> /etc/ecs/ecs.config'
02_install_ssm_agent:
  command: "#!/bin/bash\nyum -y update\nyum install -y https://amazon-ssm-eu-west-1.s3.amazonaws.com/latest/linux_amd64/amazon-ssm-agent.rpm"
```

But I would like it to produce following instead (with clean):

```yaml
01_add_instance_to_cluster:
  command: !Sub |-
    #!/bin/bash
    echo ECS_CLUSTER=${ECSCluster} >> /etc/ecs/ecs.config
02_install_ssm_agent:
  command: |-
    #!/bin/bash
    yum -y update
    yum install -y https://amazon-ssm-eu-west-1.s3.amazonaws.com/latest/linux_amd64/amazon-ssm-agent.rpm
```

Following diff might suffice:

```diff
diff --git a/cfn_flip/custom_yaml.py b/cfn_flip/custom_yaml.py
index f088d41..897ad14 100644
--- a/cfn_flip/custom_yaml.py
+++ b/cfn_flip/custom_yaml.py
@@ -141,15 +141,15 @@ def representer(dumper, data):
     elif isinstance(data, list):
         return dumper.represent_sequence(tag, data, flow_style=True)
 
-    return dumper.represent_scalar(tag, data)
+    return scalar_representer(dumper, data, tag)
 
-def scalar_representer(dumper, value):
+def scalar_representer(dumper, value, tag=TAG_STRING):
     style = None
 
     if "\n" in value:
-        style = "\""
+        style = "|"
 
-    return dumper.represent_scalar(TAG_STRING, value, style=style)
+    return dumper.represent_scalar(tag, value, style=style)
 
 # Customise our yaml
 CustomDumper.add_representer(six.text_type, scalar_representer)

```

**Comment by stilvoid** [comment]
Ok, so I see what you're getting at. The current behaviour is based on a bug report that yaml makes some assumptions about what you'd like to do with your text when you use "|".

For example, the default behaviour is to chop off trailing whitespace. I didn't want cfn-flip to make the assumption that it was safe to do this.

I think, what I'll do is to implement this style but *only* if you specify the clean flag which the docs explicitly state will perform *opinionated* changes :)

**Comment by stilvoid** [comment]
Fixed in 92dc49fd660d81b42281fde6f36dfbd1d70f6364

**Comment by azatoth** [comment]
@stilvoid I'ts still a issue where in https://github.com/awslabs/aws-cfn-template-flip/blob/master/cfn_flip/custom_yaml.py#L145 which at the moment wil represent a !Sub as a simple string (single quote) instead of a pipe block

Here I've just copied the code to show what's missing; Off course it should be generalized:

```diff
diff --git a/cfn_flip/custom_yaml.py b/cfn_flip/custom_yaml.py
index e022233..837f7a6 100644
--- a/cfn_flip/custom_yaml.py
+++ b/cfn_flip/custom_yaml.py
@@ -109,7 +109,7 @@ class ODict(collections.OrderedDict):
         items = odict_items(self.items())
         self.items = lambda: items
 
-def representer_generator(long_form=False):
+def representer_generator(long_form=False, clean_up=False):
     def representer(dumper, data):
         """
         Deal with !Ref style function format and OrderedDict
@@ -142,7 +142,12 @@ def representer_generator(long_form=False):
         elif isinstance(data, list):
             return dumper.represent_sequence(tag, data, flow_style=True)
 
-        return dumper.represent_scalar(tag, data)
+        style = None
+
+        if "\n" in data:
+            style = "|" if clean_up else "\""
+
+        return dumper.represent_scalar(tag, data, style=style)
 
     return representer
 
@@ -166,7 +171,7 @@ def dumper_generator(clean_up=False, long_form=False):
         pass
 
     Dumper.add_representer(six.text_type, scalar_representer_generator(clean_up))
-    Dumper.add_representer(collections.OrderedDict, representer_generator(long_form))
-    Dumper.add_representer(dict, representer_generator(long_form))
+    Dumper.add_representer(collections.OrderedDict, representer_generator(long_form, clean_up))
+    Dumper.add_representer(dict, representer_generator(long_form, clean_up))
 
     return Dumper
```
